recordCollection.go

```diff
@@ -25,28 +43,54 @@ func (c recordCollection) Find(needle cloudflare.DNSRecord) (int, *cloudflare.DN
 	return -1, nil
 }
 
-// Diff will find the differences between two recordCollections.
-func (c recordCollection) Diff(remote recordCollection) (recordCollection, recordCollection) {
-	localOnly := recordCollection{}
-	remoteOnly := recordCollection{}
+// Difference will find all the elements in c not present in remote [c \ remote].
+func (c recordCollection) Difference(remote recordCollection, match FilterFunc) recordCollection {
+	result := recordCollection{}
+	B := remote.Clone()
 
-	for _, l := range c {
-		n, _ := remote.Find(l)
+	for _, r := range c {
+		n, _ := B.Find(r, match)
 
 		if n < 0 {
-			localOnly = append(localOnly, l)
+			result = append(result, r)
+		} else {
+			B.Remove(n)
 		}
 	}
 
-	for _, r := range remote {
-		n, _ := c.Find(r)
+	return result
+}
 
-		if n < 0 {
-			remoteOnly = append(remoteOnly, r)
+// Intersect will find the intersection between c and remote [c ∩ remote] with
+// the caveat that the ID from c will be used in the result - while all other
+// properties will be copied from remote.
```

Comment:
It might be worth mentioning that if you use a "lenient" match function only the first match will be represented in the intersection, and it will be represented by the element from the given `remote` collection.

Eg.: Given that `f(a, b) == f(a, c) == true` then `{a, b, d}.Intersect({c,e})` yields `{c}`. Unless I'm misunderstanding something.

Reply:
I have added a comment mentioning it. It it true regardless of which match function you use, btw.